nipype/utils/filemanip.py

```diff
@@ -779,50 +813,53 @@ def read_stream(stream, logger=None, encoding=None):
 
 
     """
-    default_encoding = encoding or locale.getdefaultlocale()[1] or 'UTF-8'
+    default_encoding = encoding or locale.getdefaultlocale()[1] or "UTF-8"
     logger = logger or fmlogger
     try:
         out = stream.decode(default_encoding)
     except UnicodeDecodeError as err:
-        out = stream.decode(default_encoding, errors='replace')
-        logger.warning('Error decoding string: %s', err)
+        out = stream.decode(default_encoding, errors="replace")
+        logger.warning("Error decoding string: %s", err)
     return out.splitlines()
 
 
-def savepkl(filename, record, versioning=False):
-    pklopen = gzip.open if filename.endswith('.pklz') else open
-    with SoftFileLock('%s.lock' % filename):
-        with pklopen(filename, 'wb') as pkl_file:
-            if versioning:
-                from nipype import __version__ as version
-                metadata = json.dumps({'version': version})
+def savepkl(filename, record, versioning=False, sync=False):
+    pkl_open = gzip.open if filename.endswith(".pklz") else open
+    with pkl_open(filename, "wb") as pkl_file:
+        if versioning:
+            from nipype import __version__ as version
+
+            metadata = json.dumps({"version": version})
 
-                pkl_file.write(metadata.encode('utf-8'))
-                pkl_file.write('\n'.encode('utf-8'))
+            pkl_file.write(metadata.encode("utf-8"))
+            pkl_file.write("\n".encode("utf-8"))
 
-            pickle.dump(record, pkl_file)
+        pickle.dump(record, pkl_file)
```

Comment:
Could we do this to a bytes IO stream outside the context and then open the context and just write it?

Reply:
it is a bytes io stream and is being done with the context.

Reply:
see 827-828

Reply:
I meant something like this:

```Python
from io import BytesIO

    with BytesIO() as f:
        if versioning:
            from nipype import __version__ as version
            metadata = json.dumps({"version": version})
            f.write(metadata.encode("utf-8"))
            f.write("\n".encode("utf-8"))
        pickle.dump(record, f)
        content = f.get_value()

    pkl_open = gzip.open if filename.endswith(".pklz") else open
    with pkl_open(filename, "wb") as pkl_file:
        pkl_file.write(content)
```

Reply:
Why?

Reply:
minimize being inside the context with an open file?

Reply:
Maybe I'm missing something. Are we worried about some kind of race here? If so, why drop the lock?

Reply:
Not really. Separating steps just seemed a good idea to me in this case. We haven't tested versioning thoroughly, and doing the pickle.dump to memory seemed more efficient than all at once.

Also if we keep seeing issues, separating both will make it easier to make more targeted interventions (let's imagine we discover that gzip.open does not get along with NFS)

Reply:
(that said, I agree this is not strictly necessary for this PR, and at the risk of introducing some other bug I am happy dismissing).